### PR TITLE
Integrate mlflow tracking into layoutlm model

### PR DESCRIPTION
**Pull request by zschira**
## Background
This PR integrates mlflow into our tooling to manage the layoutlm model used for exhibit 21 tracking. There's still some refactoring/reorganization needed, but this will allow us to start collaborating on model development more easily.

## Tracking server
The tracking server lives [here](https://mlflow-ned2up6sra-uc.a.run.app) and you can login using the username: admin and password stored in gcloud secrets. There's a finetuned layoutlm model cached here that has been trained using the labeled training set with an 80/20 split of training/test data, and 1000 training steps. This model can be loaded using the method `src/mozilla_sec_eia/ex_21/extractor.py:load_model`.

I'll leave some questions in a review on next steps/how we want to organize things.

### CODE REVIEW

File: src/mozilla_sec_eia/ex_21/extractor.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+"""Fine-tune LayoutLM to extract Ex. 21 tables.
```
Comment (zschira) [inline src/mozilla_sec_eia/ex_21/extractor.py:1]: I changed this from a class to a collection of functions because it worked a bit better with mlflow that way. The layout and interface of this module can definitely be refined/changed as we start to actually use the model more.

**Comment by katie-lamb** [comment]
@zschira I like the reorg that you've done here. I don't have access to the mlflow server password in gcloud secrets, it says I need `secretmanager.versions.access` permissions, so I can't actually launch a training run right now.

File: src/mozilla_sec_eia/ex_21/rename_labeled_filings.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    unlabeled version of that document, instead the items are just
+    named with the task number from Label Studio. This script looks
+    inside the labeled JSON and gets the filename of the unlabeled
+    document. It then labels the labeled item in GCS with that
+    filename.
+    """
+    archive = GCSArchive()
+    bucket = archive._filings_bucket
+
+    labeled_bucket_name = "labeled"
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/ex_21/rename_labeled_filings.py:25]: I moved this script into the `ex_21` directory. Also, do we still need this `labeled_bucket_name` variable, or should the `GCS_LABELS_BUCKET_NAME` environment variable just include `labeled` at the end of the path?

File: src/mozilla_sec_eia/cli.py [hunk 3]
```diff
@@ -9,6 +9,8 @@
 import logging
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/cli.py:9]: The CLI isn't working for me (I'm getting "command not found" when I try to run `validate_archive` for example). Do we need a `setup.py` in the repo or is an entry point defined in some other way?

**Review by katie-lamb — COMMENTED** [review]
Besides the permissions issue for launching an MLFlow run, I was having issues with the CLI and also made a comment about the `rename_labeled_filings` script that I made.

File: src/mozilla_sec_eia/ex_21/rename_labeled_filings.py (hunk 2 above)
Comment (katie-lamb) [reply src/mozilla_sec_eia/ex_21/rename_labeled_filings.py:25]: oh okay, seems like it's better to just have the top level bucket name as the environment variable, and then specify `labeled` as the glob to get?

File: src/mozilla_sec_eia/utils/cloud.py [hunk 4]
```diff
@@ -134,45 +135,59 @@ def from_path(
...
     This class looks for several environment variables to configure
-    access to GCS resources. These can be set directly, or be in a
+    access to cloud resources. These can be set directly, or be in a
     .env file at the top level.
 
     The following variables need to be set:
 
-    GCS_BUCKET_NAME: Name of bucket where 10k filings are stored.
+    GCS_FILINGS_BUCKET_NAME: Name of bucket where 10k filings are stored.
+    GCS_LABELS_BUCKET_NAME: Name of bucket where labelled training data is stored.
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/utils/cloud.py:148]: It might be good to specify here whether the `labeled` directory should be included in this bucket name, or if it should be the name of the bucket one up from that. Although, I guess the `labeled` directory isn't really the bucket, it's just a folder within the bucket.

**Comment by katie-lamb** [comment]
Resolved permissions issues! The experiment tracking looks good and very helpful! The only metric that wasn't tracked that would be potentially helpful is the train-test split (`test_size` parameter in `ex_21.extractor.load_test_train_set`), but we probably won't mess with that too much anyways. Is it possible to add additional parameter tracking/metric tracking outside of the default settings? As a first version of this, I think it's pretty much ready!

File: src/mozilla_sec_eia/cli.py [hunk 5]
```diff
@@ -35,9 +37,19 @@ def formatter(prog) -> argparse.HelpFormatter:
...
     validate_parser.set_defaults(func=lambda: GCSArchive().validate_archive())
 
+    # Add command to fine-tune ex21 extractor
+    validate_parser = subparsers.add_parser("finetune_ex21")
+    validate_parser.add_argument("--model-output-dir", default="layoutlm_trainer")
+    validate_parser.add_argument("--test-size", default=0.2)
+    validate_parser.set_defaults(func=train_model)
+
+    # Add command to rename labeled filings on GCS
+    validate_parser = subparsers.add_parser("rename_filings")
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/cli.py:50]: Is there any reason to keep a validation parser (`validate_parser`) and make a separate parser for handling the labeled data (`label_parser`), or can you just make one generic parser (`parser`) that handles all the parsing?

File: src/mozilla_sec_eia/ex_21/rename_labeled_filings.py (hunk 2 above)
Comment (zschira) [reply src/mozilla_sec_eia/ex_21/rename_labeled_filings.py:25]: My idea here is that by pointing to the top level bucket instead of the individual folder, we have access to anything in that bucket if we/want need. It would probably be a pain if we all had our environments pointing to a single folder, then decided we actually wanted to access something from another folder in the bucket and then all have to recreate our envs.

File: src/mozilla_sec_eia/cli.py (hunk 3 above)
Comment (zschira) [reply src/mozilla_sec_eia/cli.py:9]: Yeah I just disabled the check. Not sure why it was coming up now when I haven't seen it in other repos, but shouldn't be a problem now

File: src/mozilla_sec_eia/utils/cloud.py (hunk 4 above)
Comment (zschira) [reply src/mozilla_sec_eia/utils/cloud.py]: I made this a bit more specific that it should point to the top-level bucket. Let me know if this seems clear enough!

File: src/mozilla_sec_eia/cli.py (hunk 5 above)
Comment (zschira) [reply src/mozilla_sec_eia/cli.py:50]: This is using a pattern documented [here](https://docs.python.org/3/library/argparse.html#sub-commands), which is a bit weird, but allows you to create sub-commands with their own unique arguments.

So by using multiple sub-parsers like this, we can run multiple commands with their own options all from the same top-level command like this:

`mozilla validate_archive {validate archive specific options}`
`mozilla finetune_ex21 {fine tune specific options}`

File: src/mozilla_sec_eia/utils/cloud.py (hunk 4 above)
Comment (katie-lamb) [reply src/mozilla_sec_eia/utils/cloud.py]: Looks good!

File: src/mozilla_sec_eia/cli.py (hunk 5 above)
Comment (katie-lamb) [reply src/mozilla_sec_eia/cli.py:50]: Oh got it. I guess I'm asking if `rename_filings` and `validate_archive` + `finetune_ex21` should all have use the same parser, or if there should be a different parser for `rename_filings`

**Review by katie-lamb — APPROVED** [review]
Left a comment, but looks good, approving this.
